cmd/containerboot/ingressservices.go

```diff
@@ -0,0 +1,329 @@
+// Secret. This allows the Kubernetes Operator to determine whether this proxy
+// Pod has setup firewall rules to route traffic for an ingress service.
+func (ep *ingressProxy) recordStatus(ctx context.Context, newCfg *ingressservices.Configs) error {
+	status := &ingressservices.Status{}
+	if newCfg != nil {
+		status.Configs = *newCfg
+	}
+	// Pod IPs are used to determine if recorded status applies to THIS proxy Pod.
+	status.PodIPv4 = ep.podIPv4
+	status.PodIPv6 = ep.podIPv6
+	secret, err := ep.kc.GetSecret(ctx, ep.stateSecret)
+	if err != nil {
+		return fmt.Errorf("error retrieving state Secret: %w", err)
+	}
+	bs, err := json.Marshal(status)
+	if err != nil {
+		return fmt.Errorf("error marshalling status: %w", err)
+	}
+	secret.Data[ingressservices.IngressConfigKey] = bs
+	patch := kubeclient.JSONPatch{
+		Op:    "replace",
+		Path:  fmt.Sprintf("/data/%s", ingressservices.IngressConfigKey),
+		Value: bs,
+	}
+	if err := ep.kc.JSONPatchResource(ctx, ep.stateSecret, kubeclient.TypeSecrets, []kubeclient.JSONPatch{patch}); err != nil {
+		return fmt.Errorf("error patching state Secret: %w", err)
+	}
+	return nil
+}
+
+// getRulesToAdd takes the desired firewall configuration and the recorded
+// firewall status and returns a map of missing VIPServices and rules.
+func (ep *ingressProxy) getRulesToAdd(cfgs *ingressservices.Configs, status *ingressservices.Status) map[string]ingressservices.Config {
+	if cfgs == nil {
+		return nil
+	}
+	var rulesToAdd map[string]ingressservices.Config
+	for vipSvc, wantsCfg := range *cfgs {
+		if status == nil || !ep.isCurrentStatus(status) {
+			mak.Set(&rulesToAdd, vipSvc, wantsCfg)
+			continue
+		}
+		gotCfg := status.Configs.GetConfig(vipSvc)
+		if gotCfg == nil || !reflect.DeepEqual(wantsCfg, *gotCfg) {
+			mak.Set(&rulesToAdd, vipSvc, wantsCfg)
+		}
+	}
+	return rulesToAdd
+}
+
+// getRulesToDelete takes the desired firewall configuration and the recorded
+// status and returns a map of VIPServices and rules that need to be deleted.
+func (ep *ingressProxy) getRulesToDelete(cfgs *ingressservices.Configs, status *ingressservices.Status) map[string]ingressservices.Config {
+	if status == nil || !ep.isCurrentStatus(status) {
+		return nil
+	}
+	var rulesToDelete map[string]ingressservices.Config
+	for vipSvc, gotCfg := range status.Configs {
+		if cfgs == nil {
+			mak.Set(&rulesToDelete, vipSvc, gotCfg)
+			continue
+		}
+		wantsCfg := cfgs.GetConfig(vipSvc)
+		if wantsCfg != nil && reflect.DeepEqual(*wantsCfg, gotCfg) {
+			continue
+		}
+		mak.Set(&rulesToDelete, vipSvc, gotCfg)
+	}
+	return rulesToDelete
+}
+
+// ensureIngressRulesAdded takes a map of VIPServices and rules and ensures that the firewall rules are added.
+func ensureIngressRulesAdded(cfgs map[string]ingressservices.Config, nfr linuxfw.NetfilterRunner) error {
+	for serviceName, cfg := range cfgs {
+		f := func(svcName string, vipIP, clusterIP netip.Addr) error {
```

Comment:
nit: In this function and the function below it's a bit odd/buggy to define the function inside the loop and use a capture for `serviceName` but also pass in an argument `svcName`. Also I'm not sure if the go compiler is smart enough to optimise allocations for this on its own but maybe move the function outside the loop?

Reply:
Oh I didn't spot the serviceName issue. Pulled this out into a helper